sdk/src/env.rs

```diff
@@ -0,0 +1,74 @@
+use cap_sdk_core::{Index, RootBucket, Router};
+use futures::future::LocalBoxFuture;
+use ic_kit::ic::{get_maybe, store};
+use std::cell::RefCell;
+
+/// Contains data about the cap environment.
+#[derive(Clone)]
+pub struct CapEnv {
+    pub(crate) root: RootBucket,
+    pub(crate) router: Router,
+}
+
+thread_local! {
+    pub(crate) static FUTURES: RefCell<Vec<LocalBoxFuture<'static, ()>>> = RefCell::new(vec![]);
+}
+
+impl CapEnv {
+    /// Creates a new [`CapEnv`] with the index canister's [`Principal`] set to `index`.
+    pub(crate) fn create(root: RootBucket, router: Router) -> Self {
+        CapEnv { root, router }
+    }
+
+    /// Stores the [`CapEnv`] in the canister.
+    pub(crate) fn store(&self) {
+        store(self.clone());
+    }
+
+    pub(crate) fn get<'a>() -> &'a Self {
+        if let Some(data) = get_maybe::<CapEnv>() {
+            data
+        } else {
+            panic!("No context created.");
+        }
+    }
+
+    pub(crate) fn index(&self) -> Index {
+        self.router.into()
+    }
+
+    pub(crate) async fn await_futures() {
+        let futures = FUTURES.with(|futures| {
+            let mut inner = futures.take();
+
+            inner.drain(0..inner.len()).collect::<Vec<_>>()
+        });
+
+        for future in futures {
+            future.await;
+        }
+    }
```

Comment:
Okay this is problematic... Imagine a contract is live and calls `handshake` during `init`, we're deferring the execution until an `insert` call (which invokes this `await_futures`)

Now two concurrent calls come in to the system both invoking an insert, the goal is to have a ready `Env` when we're running the actual logic for insertion.

Let's model our flow which we will use for this proof.
```
global is_ready = false;
global task_taken = false;

async def handshake():
  // an async call.
  is_ready = true;

async def await_futures():
  if task_taken:
    log("Skipping task");
    return;
  log("Running task");
  task_taken = true;  
  handshake().await;
  log("Task completed");

async def insert():
  log("Insert invoked");
  await_futures().await;
  log("Await futures returned");
  assert(is_ready);
  log("Env is ready");
```

So let's first see what happens when there is only one concurrent call to `insert()`

```
> insert()
- Insert invoked
- Running task
- [we're awaiting]
- Task completed
- Await futures returned
- Env is ready
```

Fine, it worked, now let's do it for two concurrent calls to insert. (For better understanding I've prefixed messages from the first insert call with `-` and the second one with `#`)

```
> insert()
> insert()

- Insert invoked
- Running Task
- [we're awaiting]
# Insert invoked
# Skipping task
# Await futures returned
# Assertion failed ***
- Task completed
- Await futures returned
- Env is ready
```

So as you can see any call to insert will fail the assertion while handshake is not actually completed.

# Solution

let's copy this from [here](https://docs.rs/futures/0.3.17/futures/task/struct.AtomicWaker.html)

```
struct Inner {
    waker: AtomicWaker,
    set: AtomicBool,
}

#[derive(Clone)]
struct Flag(Arc<Inner>);

impl Flag {
    pub fn new() -> Self {
        Self(Arc::new(Inner {
            waker: AtomicWaker::new(),
            set: AtomicBool::new(false),
        }))
    }

    pub fn signal(&self) {
        self.0.set.store(true, Relaxed);
        self.0.waker.wake();
    }
}

impl Future for Flag {
    type Output = ();

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
        // quick check to avoid registration if already done.
        if self.0.set.load(Relaxed) {
            return Poll::Ready(());
        }

        self.0.waker.register(cx.waker());

        // Need to check condition **after** `register` to avoid a race
        // condition that would result in lost notifications.
        if self.0.set.load(Relaxed) {
            Poll::Ready(())
        } else {
            Poll::Pending
        }
    }
}
```

Now let's change the code to this:
```suggestion
    pub(crate) async fn await_futures() {
        let futures = FUTURES.with(|futures| {
            let mut inner = futures.take();

            inner.drain(0..inner.len()).collect::<Vec<_>>()
        });

        let flag = Flag::new();
        let f2 = flag.clone();
 
        let closure = async {
          flag.await;
        };

        CapEnv::insert_future(Box::pin(closure))

        for future in futures {
            future.await;
        }

        f2.signal();
    }
```